R/derive_param_confirmed_resp.R

```diff
@@ -0,0 +1,386 @@
+#'   ) %>%
+#'     mutate(PARAMCD = "PD"))
+#'
+#' adrs <- bind_rows(ovr_obs, pd_obs) %>%
+#'   mutate(
+#'     ADT = lubridate::ymd(ADTC),
+#'     STUDYID = "XX1234"
+#'   ) %>%
+#'   select(-ADTC)
+#'
+#' pd_date <- date_source(
+#'   dataset_name = "adrs",
+#'   date = ADT,
+#'   filter = PARAMCD == "PD"
+#' )
+#'
+#' # Derive confirmed best overall response parameter
+#' derive_param_confirmed_resp(
+#'   adrs,
+#'   dataset_adsl = adsl,
+#'   filter_source = PARAMCD == "OVR",
+#'   source_pd = pd_date,
+#'   source_datasets = list(adrs = adrs),
+#'   ref_confirm = 28,
+#'   set_values_to = vars(
+#'     PARAMCD = "CRSP",
+#'     PARAM = "Confirmed Response by Investigator"
+#'   )
+#' ) %>%
+#'   filter(PARAMCD == "CRSP")
+#'
+#' # Derive confirmed best overall response parameter (accepting SD for PR and two NEs)
+#' derive_param_confirmed_resp(
+#'   adrs,
+#'   dataset_adsl = adsl,
+#'   filter_source = PARAMCD == "OVR",
+#'   source_pd = pd_date,
+#'   source_datasets = list(adrs = adrs),
+#'   ref_confirm = 28,
+#'   max_nr_ne = 2,
+#'   accept_sd = TRUE,
+#'   set_values_to = vars(
+#'     PARAMCD = "CRSP",
+#'     PARAM = "Confirmed Response by Investigator"
+#'   )
+#' ) %>%
+#'   filter(PARAMCD == "CRSP")
+derive_param_confirmed_resp <- function(dataset,
+                                        dataset_adsl,
+                                        filter_source,
+                                        source_pd = NULL,
+                                        source_datasets,
+                                        ref_confirm,
+                                        max_nr_ne = 1,
+                                        accept_sd = FALSE,
+                                        aval_fun = yn_to_numeric,
+                                        set_values_to,
+                                        subject_keys = vars(STUDYID, USUBJID)) {
+  # Check input parameters
+  filter_source <- assert_filter_cond(enquo(filter_source))
+  assert_s3_class(source_pd, "date_source", optional = TRUE)
+  assert_list_of(source_datasets, "data.frame")
+  assert_integer_scalar(ref_confirm, subset = "non-negative")
+  assert_integer_scalar(max_nr_ne, subset = "non-negative")
+  assert_logical_scalar(accept_sd)
+  assert_function(aval_fun)
+  assert_varval_list(set_values_to, required_elements = "PARAMCD")
+  assert_vars(subject_keys)
+  assert_data_frame(
+    dataset,
+    required_vars = quo_c(subject_keys, vars(PARAMCD, ADT, AVALC))
+  )
+  assert_data_frame(dataset_adsl, required_vars = subject_keys)
+  if (!is.null(dataset)) {
+    assert_param_does_not_exist(dataset, quo_get_expr(set_values_to$PARAMCD))
+  }
+
+  # Restrict input dataset
+  source_data <- dataset %>%
+    filter_pd(
+      filter = !!filter_source,
+      source_pd = source_pd,
+      source_datasets = source_datasets
+    )
+
+  # Check for invalid AVALC values
+  resp_vals <- source_data$AVALC
+  valid_vals <- c("CR", "PR", "SD", "NON-CR/NON-PD", "PD", "NE", "ND")
```

Comment:
some vendors also use 'NA' instead of ND ( no disease), can we add to the list in case?

Reply:
@amitjaingsk  we'll take care of this later via https://github.com/pharmaverse/admiralonco/issues/45 - i added your comment there